File: fl_backend/app/api/zustellung/schemas.py
from typing import Literal

from pydantic import BaseModel, ConfigDict

# The application's slice is where the delivery vocabulary was first written, and a second spelling
# of the six states would let two homes disagree about what a bounce is called.
from app.api.bewerbungen.schemas import CustomNachrichtId, CustomZustellgrund, CustomZustellzeitpunkt, FLBewerbungZustellEreignis
from app.shared.schemas.custom import CustomObjectId
from app.shared.schemas.responses import BaseAPIResponse

# A member joins this set in the commit that gives it an `app/api/zustellung/services.py ::
# ZIEL_PFADE` row and its own validator block: one arriving without either names a kind every
# write reaches nothing through.
FLZustellungZiel = Literal["schiedsrichter", "einladung", "registrierung"]


class _ZielMeldung(BaseModel):
    """What every write names: the kind of record, the row itself, and when."""

    model_config = ConfigDict(extra="forbid")

    ziel: FLZustellungZiel
    # Beside the kind, which names a population: a report carrying `ziel` alone reaches whichever
    # row a later reader guesses at. In the BODY, as the application's twin keeps its own id there,
    # the edge logging a path (`docs/logging/spec.md :: L9`).
    ziel_id: CustomObjectId
    am: CustomZustellzeitpunkt


class _ZielZustellungPayload(_ZielMeldung):
    """What the two writes about a minted message name, the message among them."""

    nachricht_id: CustomNachrichtId


class FLZustellungAngenommenPayload(_ZielZustellungPayload):
    """One message the provider accepted for this record. `stand` is not a key: this endpoint records `angenommen` and no other state."""


class FLZustellungAbgewiesenPayload(_ZielMeldung):
    """A send the provider refused outright. No `nachricht_id`: nothing was minted, so there is no message for a later event to join."""

    # Required as a KEY and null where the refusal names no token, as the event's twin requires it:
    # an omitted key would read as a client that forgot it rather than a refusal carrying none.
    grund: CustomZustellgrund


class FLZustellungEreignisPayload(_ZielZustellungPayload):
    """One delivery event about a message this record was sent."""

    stand: FLBewerbungZustellEreignis
    # Required as a KEY and null where the event carries none, as the application's twin requires
    # it: an omitted key would read as a client that forgot it rather than a bounce with no token.
    grund: CustomZustellgrund


class FLZustellungResponse(BaseAPIResponse):
    """Whether the write reached the record; false where the report is about a message the record does not hold.

    False rather than a refusal: the provider repeats a settled event for hours if anything but a
    success answers it.
    """

    angewendet: bool

File: fl_frontend/src/features/zustellung/schemas.ts
import z from "zod";

import { BaseAPIResponseSchema } from "@/core/schemas";
import {
  einzeiligerName,
  FLBewerbungZustellstandSchema,
  ZUSTELLUNG_GRUND_MAX_LENGTH,
  ZUSTELLUNG_NACHRICHT_ID_MAX_LENGTH,
  ZUSTELLUNG_ZEITPUNKT_MAX_LENGTH,
} from "@/features/bewerbungen/schemas";
import { CustomObjectIdStringSchema } from "@/shared/schemas";

/**
 * Mirrors `fl_backend/app/api/zustellung/schemas.py :: FLZustellungZiel` — which kind of record a
 * report is about. Closed on both sides, so a kind the register has no home for is refused here
 * rather than dispatched to nothing.
 */
export const FLZustellungZielSchema = z.enum(["schiedsrichter", "einladung"], { error: "Diesen Empfängertyp gibt es nicht." });
export type FLZustellungZiel = z.infer<typeof FLZustellungZielSchema>;

/**
 * What both generic delivery writes name, mirroring the one private base they share on the backend.
 * Spelled once for the reason the base exists: two copies would let the acceptance and the event
 * judge one provider's message differently.
 */
const zielZustellungMeldungFields = {
  ziel: FLZustellungZielSchema,
  // Beside the kind, which names a population rather than a row: an event carrying the kind alone
  // would be applied to whichever record a later reader guessed at.
  ziel_id: CustomObjectIdStringSchema,
  nachricht_id: z
    .string()
    .trim()
    // German though no reader meets one: this server composes both writes, and a refusal reaches the
    // log rather than a box. `fl_frontend/src/core/schemaGerman.test.ts` holds every payload to it.
    .min(1, { error: "Diese Meldung nennt keine Nachricht." })
    .max(ZUSTELLUNG_NACHRICHT_ID_MAX_LENGTH, { error: "Diese Nachrichten-ID ist zu lang." }),
  am: z
    .string()
    .trim()
    .min(1, { error: "Diese Meldung nennt keinen Zeitpunkt." })
    .max(ZUSTELLUNG_ZEITPUNKT_MAX_LENGTH, { error: "Dieser Zeitpunkt ist zu lang." }),
};

/** Mirrors `FLZustellungAngenommenPayload` — one message the provider took. No `stand`: that endpoint records `angenommen` and refuses anything else. */
export const FLZustellungAngenommenPayloadSchema = z.object(zielZustellungMeldungFields);
export type FLZustellungAngenommenPayload = z.infer<typeof FLZustellungAngenommenPayloadSchema>;

/** Mirrors `FLZustellungEreignisPayload` — one event the provider sent about a message already recorded. */
export const FLZustellungEreignisPayloadSchema = z.object({
  ...zielZustellungMeldungFields,
  // Without `angenommen`, which the acceptance endpoint alone writes: a webhook composing one would
  // be refused 422 and retried for thirty-two hours over a state no event of the provider's carries.
  stand: FLBewerbungZustellstandSchema.exclude(["angenommen"]),
  // No floor beside the ceiling: the endpoint takes an empty reason and a `null` alike, a bounce
  // carrying no token still being a fact about the mailbox. The class is screened as the endpoint
  // screens it.
  grund: einzeiligerName(z.string().trim().max(ZUSTELLUNG_GRUND_MAX_LENGTH, { error: "Dieser Grund ist zu lang." }), "Der Grund").nullable(),
});
export type FLZustellungEreignisPayload = z.infer<typeof FLZustellungEreignisPayloadSchema>;

/** Whether the write reached the record. `false` is a superseded message or an event already answered, and is not a refusal. */
export const FLZustellungResponseSchema = BaseAPIResponseSchema.extend({
  angewendet: z.boolean(),
});
export type FLZustellungResponse = z.infer<typeof FLZustellungResponseSchema>;
